light-blockchain/tests/push.rs

```diff
@@ -410,12 +412,214 @@ fn it_can_rebranch_skip_block() {
     let fork1 = temp_producer2.next_block(vec![], true);
 
     let inferior2 = temp_producer1.next_block(vec![], false);
-    temp_producer2.next_block(vec![], false);
+    let fork2 = temp_producer2.next_block(vec![], false);
+
+    // Check that producer 2 ignores inferior chain.
+    assert_eq!(temp_producer2.push(inferior1), Ok(PushResult::Ignored));
+    assert_eq!(temp_producer2.push(inferior2), Ok(PushResult::Ignored));
+
+    // Check that producer 1 rebranches.
+    assert_eq!(temp_producer1.push(fork1), Ok(PushResult::Rebranched));
+    assert_eq!(temp_producer1.push(fork2), Ok(PushResult::Extended));
+
+    // Case 2: not obvious rebranch rebranch (number denotes accumulated skip block)
+    // ... - [1] - [1] - [2]
+    //          \- [2] - [2]
+    let block = temp_producer1.next_block(vec![], false);
+    temp_producer2.push(block).unwrap();
+
+    let inferior1 = temp_producer1.next_block(vec![], false);
+    let fork1 = temp_producer2.next_block(vec![], true);
+
+    let inferior2 = temp_producer1.next_block(vec![], true);
+    let fork2 = temp_producer2.next_block(vec![], false);
 
     // Check that producer 2 ignores inferior chain.
     assert_eq!(temp_producer2.push(inferior1), Ok(PushResult::Ignored));
     assert_eq!(temp_producer2.push(inferior2), Ok(PushResult::Ignored));
 
     // Check that producer 1 rebranches.
     assert_eq!(temp_producer1.push(fork1), Ok(PushResult::Rebranched));
+    assert_eq!(temp_producer1.push(fork2), Ok(PushResult::Extended));
+}
+
+#[test]
+fn micro_block_works_after_macro_block() {
+    let temp_producer = TemporaryLightBlockProducer::new();
+
+    // apply an entire batch including macro block on skip_block/round_number zero
+    for _ in 0..Policy::blocks_per_batch() {
+        temp_producer.next_block(vec![], false);
+    }
+    // make sure we are at the beginning of the batch and all block were applied
+    assert_eq!(
+        temp_producer.blockchain.read().block_number(),
+        Policy::blocks_per_batch()
+    );
+
+    // Test if a micro block can be rebranched immediately after
+    // a round_number 0 macro block
+
+    // Create a couple of skip blocks
+    let block = temp_producer.next_block_no_push(vec![], true);
+    let rebranch = temp_producer.next_block_no_push(vec![], true);
+    // Push first skip block
+    temp_producer.push(block).unwrap();
+    // Make sure this was an extend
+    assert_eq!(
+        temp_producer.blockchain.read().block_number(),
+        Policy::blocks_per_batch() + 1
+    );
+    // And rebranch it to block the chain with only one skip block
+    temp_producer.push(rebranch).unwrap();
+    // Make sure this was a rebranch
+    assert_eq!(
+        temp_producer.blockchain.read().block_number(),
+        Policy::blocks_per_batch() + 1
+    );
+
+    // Apply the rest of the batch including macro block on skip_block/round_number one
```

Comment:
Mmm, this doesn't sound good. What is a macro block on skip block one? Maybe this should only say `round_number`?